go/vt/mysqlctl/mysqlshellbackupengine.go

```diff
@@ -164,41 +165,38 @@ func (be *MySQLShellBackupEngine) ExecuteBackup(ctx context.Context, params Back
 		return BackupUnusable, vterrors.Wrap(err, "failed to fetch position")
 	}
 
-	cmd := exec.CommandContext(ctx, mysqlShellBackupBinaryName, args...)
+	cmd := exec.CommandContext(ctx, be.binaryName, args...)
 
 	params.Logger.Infof("running %s", cmd.String())
 
-	cmdOut, err := cmd.StdoutPipe()
-	if err != nil {
-		return BackupUnusable, vterrors.Wrap(err, "cannot create stdout pipe")
-	}
-	cmdOriginalErr, err := cmd.StderrPipe()
-	if err != nil {
-		return BackupUnusable, vterrors.Wrap(err, "cannot create stderr pipe")
-	}
-	if err := cmd.Start(); err != nil {
-		return BackupUnusable, vterrors.Wrap(err, "can't start mysqlshell")
-	}
+	stdoutReader, stdoutWriter := io.Pipe()
+	stderrReader, stderrWriter := io.Pipe()
+	lockWaiterReader, lockWaiterWriter := io.Pipe()
 
-	pipeReader, pipeWriter := io.Pipe()
-	cmdErr := io.TeeReader(cmdOriginalErr, pipeWriter)
+	cmd.Stdout = stdoutWriter
+	cmd.Stderr = stderrWriter
+	combinedErr := io.TeeReader(stderrReader, lockWaiterWriter)
 
 	cmdWg := &sync.WaitGroup{}
 	cmdWg.Add(3)
-	go releaseReadLock(ctx, pipeReader, params, cmdWg, lockAcquired)
-	go scanLinesToLogger(mysqlShellBackupEngineName+" stdout", cmdOut, params.Logger, cmdWg.Done)
-	go scanLinesToLogger(mysqlShellBackupEngineName+" stderr", cmdErr, params.Logger, cmdWg.Done)
+	go releaseReadLock(ctx, lockWaiterReader, params, cmdWg, lockAcquired)
+	go scanLinesToLogger(mysqlShellBackupEngineName+" stdout", stdoutReader, params.Logger, cmdWg.Done)
+	go scanLinesToLogger(mysqlShellBackupEngineName+" stderr", combinedErr, params.Logger, cmdWg.Done)
+
+	// we run the command, wait for it to complete and close all pipes so the goroutines can complete on their own.
+	// after that we can process if an error has happened or not.
+	err = cmd.Run()
+
+	time.Sleep(time.Millisecond * 100) // give the goroutines some time to read any remaining logs before closing pipes.
```

Comment:
I can consistently run this from the `main` branch and hit that edge case:
```
$ go test -run=TestMySQLShellBackupEngine_ExecuteBackup_ReleaseLock -race -count=1000
--- FAIL: TestMySQLShellBackupEngine_ExecuteBackup_ReleaseLock (0.01s)
    --- FAIL: TestMySQLShellBackupEngine_ExecuteBackup_ReleaseLock/lock_released_if_we_see_the_mysqlsh_lock_being_acquired (0.00s)
        mysqlshellbackupengine_test.go:379:
            	Error Trace:	/home/rrangel/vitess_rvrangel/go/vt/mysqlctl/mysqlshellbackupengine_test.go:379
            	Error:      	"I1126 15:18:30.734269 mysqlshellbackupengine.go:117] Starting ExecuteBackup in test\nI1126 15:18:30.734448 mysqlshellbackupengine.go:150] acquiring a global read lock before fetching the executed GTID sets\nI1126 15:18:30.734470 mysqlshellbackupengine.go:169] running /tmp/TestMySQLShellBackupEngine_ExecuteBackup_ReleaseLock892506315/001/test.sh --defaults-file=/dev/null --js -h localhost -e util.dumpInstance(\"logical/tmp/TestMySQLShellBackupEngine_ExecuteBackup_ReleaseLocklock_released_if_we_see_the_mysqlsh_lock_being_acquired531729948/001\", {\"threads\": 4})\nW1126 15:18:30.737258 backup.go:524] error scanning lines from mysqlshell stdout: read |0: file already closed\nW1126 15:18:30.737293 backup.go:524] error scanning lines from mysqlshell stderr: read |0: file already closed\nE1126 15:18:30.737314 mysqlshellbackupengine.go:534] could not release global lock earlier\nI1126 15:18:30.737335 mysqlshellbackupengine.go:203] Writing backup MANIFEST\nI1126 15:18:30.737503 mysqlshellbackupengine.go:243] Backup completed\nI1126 15:18:30.737512 xtrabackupengine.go:164] Closing backup file MANIFEST\n" does not contain "global read lock released after"
            	Test:       	TestMySQLShellBackupEngine_ExecuteBackup_ReleaseLock/lock_released_if_we_see_the_mysqlsh_lock_being_acquired
            	Messages:   	failed to release the global lock after mysqlsh
FAIL
cleaning up "/tmp/50bd7cc81b5edaf5bdb887a7c4ab3ba5.dat"
exit status 1
FAIL	vitess.io/vitess/go/vt/mysqlctl	12.238s
```

with this PR's branch, I reduced the sleep time to `10ms` to allow me to runs it thousands of times and haven't seen the issue once the Sleep has been added:
```
$ go test -run=TestMySQLShellBackupEngine_ExecuteBackup_ReleaseLock -race -count=10000
PASS
cleaning up "/tmp/50bd7cc81b5edaf5bdb887a7c4ab3ba5.dat"
ok  	vitess.io/vitess/go/vt/mysqlctl	448.584s
```
I have tested this with `10ms` on thousands of tests and didn't see it fail, `100ms` should be safer to avoid the flaky test.

Reply:
I don't think we should use sleep, why wouldn't it work to wait for the group then? Then the goroutines are done and would have read it all? 

In general, a sleep is kinda an anti pattern for concurrent work. If it's needed, there's usually a better way to solve the problem.

Reply:
the problem is that closing it too fast once the program finishes doesn't give enough time for `bufio` to Scan the last line, so the test doesn't see the lock being released. and if we close it immediately, then there is this race between Scan() trying to read it while the file is closed and it doesn't report the last line.

In the real world this should never really happen as the line we are looking for is printed in the beginning, but it shows up immediately on the mocked test. I can move the sleep to the inside the mocked program, which would avoid having this in the code

Reply:
Going to merge this since it affects CI so significantly. I still think there's something here fundamentally broken and it needs some rework to remove the actual race, but not going to hold this back here.